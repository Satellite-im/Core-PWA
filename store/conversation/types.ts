export enum ConversationConnection {
  CONNECTED = 'CONNECTED',
  CONNECTING = 'CONNECTING',
  DISCONNECTED = 'DISCONNECTED',
  DISCONNECTING = 'DISCONNECTING',
  ERROR = 'ERROR',
}

export enum ConversationActivity {
  ACTIVE = 'ACTIVE',
  INACTIVE = 'INACTIVE',
  TYPING = 'TYPING',
  NOT_TYPING = 'NOT_TYPING',
}

export type ConversationParticipant = {
  peerId?: string
  address: string
  name?: string // will be undefined for self
  profilePicture?: string
  state?: ConversationConnection
  activity?: ConversationActivity
  updatedAt?: number
  textilePubkey?: string // will be undefined for self
}

export type ConversationState = {
  id: string
  type: 'friend' | 'group'
  calling: boolean
  participants: Array<ConversationParticipant>
}
