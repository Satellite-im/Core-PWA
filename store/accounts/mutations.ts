import { AccountsState, RegistrationStatus } from './types'
import { User } from '~/libraries/Iridium/users/types'

const mutations = {
  setPin(state: AccountsState, pin: string) {
    state.pin = pin
  },
  setStorePin(state: AccountsState, storePin: boolean) {
    state.storePin = storePin
  },
  setRegistry(state: AccountsState, registry: boolean) {
    state.registry = registry
  },
  setPinHash(state: AccountsState, pinHash: string) {
    state.pinHash = pinHash
  },
  unlock(state: AccountsState, pin: string) {
    state.locked = false
    state.pin = pin
  },
  lock(state: AccountsState) {
    state.locked = true
  },
  setEncryptedPhrase(state: AccountsState, encryptedPhrase: string) {
    state.encryptedPhrase = encryptedPhrase
  },
  setPhrase(state: AccountsState, decryptedPhrase: string) {
    state.phrase = decryptedPhrase
  },
  setActiveAccount(state: AccountsState, activeAccountPubkey: string) {
    state.active = activeAccountPubkey
    state.initialized = true
  },
  setProfilePicture(state: AccountsState, image: string) {
    if (state.details) {
      state.details.photoHash = image
    }
  },
  setUserDetails(state: AccountsState, details: User) {
    state.details = {
      did: details.did,
      name: details.name,
      status: details.status,
      photoHash: details.photoHash,
    }
    state.welcomePopup = true
  },
  dismissWelcome(state: AccountsState) {
    state.welcomePopup = false
  },
  setUserPeerId(state: AccountsState, peerId: string) {
    if (state.details) {
      state.details = {
        ...state.details,
        peerId,
      }
    }
  },
  setRegistrationStatus(
    state: AccountsState,
    registrationStates: RegistrationStatus,
  ) {
    state.registrationStatus = registrationStates
  },
  setLastVisited(state: AccountsState, lastVisited: string) {
    state.lastVisited = lastVisited
  },
  setAdapter(state: AccountsState, adapter: string) {
    state.adapter = adapter
  },
  setEntropy(state: AccountsState, entropyMessage: string) {
    state.entropyMessage = entropyMessage
  },
}

export default mutations
