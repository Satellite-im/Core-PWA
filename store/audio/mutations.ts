import { AudioState } from './types'

const mutations = {
  mute(state: AudioState) {
    state.muted = !state.muted
  },
  toggleMute(state: AudioState) {
    state.muted = !state.muted
  },
  deafen(state: AudioState) {
    const isDeafened = state.deafened
    state.previousVolume = isDeafened ? state.previousVolume : state.volume
    state.volume = isDeafened ? state.previousVolume : 0
    state.deafened = !isDeafened
  },
  setVolume(state: AudioState, volume: Number) {
    state.previousVolume = volume
    state.volume = volume
  },
  setSystemVolume(state: AudioState, systemVolume: Number) {
    state.sounds.system = systemVolume
  },
  setInputVolume(state: AudioState, inputVolume: Number) {
    state.inputVolume = inputVolume
  },
  setSoundLevel(
    state: AudioState,
    sound: {
      inboundMedia: Number
      outboundMedia: Number
      system: Number
    },
  ) {
    state.sounds = sound
  },
}

export default mutations
