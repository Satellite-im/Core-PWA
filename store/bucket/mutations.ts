import { BucketState } from './types'
import { FilSystem } from '~/libraries/Files/FilSystem'

const mutations = {
  bucketInitialized(state: BucketState, status: boolean) {
    state.initialized = status
  },
  updateFileSystem(state: BucketState, fs: FilSystem) {
    state.fileSystem = fs
  },
  openDirectory(state: BucketState, name: string) {
    state.fileSystem.openDirectory(name)
  },
  createDirectory(state: BucketState, name: string) {
    state.fileSystem.createDirectory(name)
  },
  goBackToDirectory(state: BucketState, name: string) {
    state.fileSystem.goBackToDirectory(name)
  },
  createFile(state: BucketState, name: string) {
    state.fileSystem.createFile(name)
  },
}

export default mutations
