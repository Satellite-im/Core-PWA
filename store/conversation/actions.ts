import { PublicKey } from '@solana/web3.js'
import { keys } from 'libp2p-crypto'
import { createFromPubKey } from 'peer-id'
import {
  ConversationActivity,
  ConversationParticipant,
  ConversationState,
} from './types'
import { ActionsArguments } from '~/types/store/store'

const actions = {
  /**
   * @method initialize
   * @description Initialize the conversation store
   * @example
   * store.dispatch('conversation/initialize');
   */
  initialize({ commit }: ActionsArguments<ConversationState>) {
    commit('setConversation', {
      id: '',
      type: 'friend',
      calling: false,
      participants: [],
    })
  },
  /**
   * @method setConversation
   * @description Set the current conversation
   * @example
   * store.dispatch('conversation/setConversation', conversation);
   */
  setConversation(
    { commit }: ActionsArguments<ConversationState>,
    payload: {
      id: string
      type: 'friend' | 'group'
      calling?: boolean
      participants: Array<ConversationParticipant>
    },
  ) {
    commit('setConversation', payload)
  },
  /**
   * @method setCalling
   * @description Set the calling state of the conversation
   * @example
   * store.dispatch('conversation/setCalling', true);
   */
  setCalling(
    { commit }: ActionsArguments<ConversationState>,
    calling: boolean,
  ) {
    commit('setCalling', calling)
  },
  /**
   * @method addParticipant
   * @description Add a participant to the conversation
   * @example
   * store.dispatch('conversation/addParticipant', participant);
   */
  async addParticipant(
    { commit, state, rootState }: ActionsArguments<ConversationState>,
    address: string,
  ) {
    const friend = rootState.friends.all.find((fr) => fr.address === address)
    const participant = state.participants?.find((p) => p.address === address)

    let peerId = friend?.peerId || participant?.peerId
    if (!peerId) {
      peerId = (
        await createFromPubKey(
          keys.supportedKeys.ed25519.unmarshalEd25519PublicKey(
            new PublicKey(address).toBytes(),
          ).bytes,
        )
      ).toB58String()
    }

    const participantDescriptor = {
      address,
      peerId,
      textilePubkey: friend?.textilePubkey || participant?.textilePubkey,
      name: friend?.name || participant?.name,
      state: friend
        ? friend.state === 'online'
          ? 'CONNECTED'
          : 'DISCONNECTED'
        : participant?.state || 'DISCONNECTED',
      profilePicture: participant?.profilePicture,
      activity: ConversationActivity.NOT_TYPING,
    }

    commit(
      participant ? 'updateParticipant' : 'addParticipant',
      participantDescriptor,
    )
  },
  /**
   * @method addParticipants
   * @description Add multiple participants to the conversation
   * @param participants
   * @example
   * store.dispatch('conversation/addParticipants', participants);
   */
  addParticipants(
    action: ActionsArguments<ConversationState>,
    participants: string[],
  ) {
    participants.forEach((address) => {
      action.dispatch('addParticipant', address)
    })
  },
}
export default actions
