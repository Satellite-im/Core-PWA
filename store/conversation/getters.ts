import {
  ConversationActivity,
  ConversationConnection,
  ConversationParticipant,
  ConversationState,
} from '~/store/conversation/types'
import { RootState } from '~/types/store/store'
import { Friend } from '~/types/ui/friends'
import { Group } from '~/store/groups/types'

const getters = {
  /**
   * @method otherParticipants
   * @description get participants other than yourself
   */
  otherParticipants: (state: ConversationState): ConversationParticipant[] => {
    return state.participants.filter((participant) => participant.name != null)
  },
  /**
   * @method onlineParticipants
   * @description array of online participants other than yourself
   */
  onlineParticipants: (
    state: ConversationState,
    getters: { otherParticipants: ConversationParticipant[] },
  ): ConversationParticipant[] => {
    return getters.otherParticipants.filter(
      (participant) => participant.state === ConversationConnection.CONNECTED,
    )
  },
  /**
   * @method typingParticipants
   * @description array of online participants other than yourself
   */
  typingParticipants: (
    state: ConversationState,
    getters: { otherParticipants: ConversationParticipant[] },
  ): ConversationParticipant[] => {
    return getters.otherParticipants.filter(
      (participant) => participant.activity === ConversationActivity.TYPING,
    )
  },
  /**
   * @method isGroup
   * @description is current recipient group
   */
  isGroup(state: ConversationState) {
    return state.type === 'group'
  },
  /**
   * @method recipient
   * @description current recipient full info
   */
  recipient: (
    state: ConversationState,
    getters: any,
    rootState: RootState,
  ): Friend | Group | undefined => {
    return getters.isGroup
      ? rootState.groups.all.find((group: Group) => group.id === state.id)
      : rootState.friends.all.find(
          (friend: Friend) => friend.peerId === state.id,
        )
  },
}

export default getters
