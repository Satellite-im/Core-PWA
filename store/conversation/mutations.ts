import { ConversationParticipant, ConversationState } from './types'

const mutations = {
  setConversation(
    state: ConversationState,
    payload: {
      id: string
      type: 'friend' | 'group'
      participants: Array<ConversationParticipant>
      calling?: boolean
    },
  ) {
    state.id = payload.id
    state.type = payload.type
    state.calling = !!payload.calling
    state.participants = payload.participants || []
  },
  setCalling(state: ConversationState, active: boolean) {
    state.calling = active
  },
  setParticipants(
    state: ConversationState,
    participants: Array<ConversationParticipant>,
  ) {
    state.participants = participants || []
  },
  addParticipant(
    state: ConversationState,
    participant: ConversationParticipant,
  ) {
    state.participants = [...state.participants, participant]
  },
  updateParticipant(
    state: ConversationState,
    participant: ConversationParticipant,
  ) {
    const prev = state.participants.find(
      (p) =>
        p.address === participant.address ||
        (p.peerId && p.peerId === participant.peerId),
    )
    if (!prev) {
      return
    }
    const next = { ...prev, ...participant }

    // avoid updating the participant unless we have to (video blink issue from peer announce)
    if (JSON.stringify(next) === JSON.stringify(prev)) {
      return
    }

    state.participants = state.participants.map((p) => {
      if (
        p.address === participant.address ||
        (p.peerId && p.peerId === participant.peerId)
      ) {
        return next
      }
      return p
    })
  },
}
export default mutations
