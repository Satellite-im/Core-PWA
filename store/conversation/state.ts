import { ConversationState } from './types'

export default function InitialConversationState(): ConversationState {
  return {
    id: '',
    type: 'friend',
    calling: false,
    participants: [],
  }
}
