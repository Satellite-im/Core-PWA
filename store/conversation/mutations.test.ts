import {
  ConversationState,
  ConversationConnection,
  ConversationActivity,
} from './types'
import * as mutations from '~/store/conversation/mutations'

describe('mutations.default.setConversation', () => {
  test('0', () => {
    const result: any = mutations.default.setConversation(
      { calling: 'error\n', participants: 'DELETE FROM ' },
      {
        id: '03ea49f8-1d96-4cd0-b279-0684e3eec3a9',
        type: 'array',
        calling: 'ValueError',
        participants: 'SELECT * FROM %s LIMIT 1',
      },
    )
    expect(result).toMatchSnapshot()
  })

  test('1', () => {
    const result: any = mutations.default.setConversation(
      { calling: 'Message box: foo; bar\n', participants: 'DELETE FROM ' },
      {
        id: 'a85a8e6b-348b-4011-a1ec-1e78e9620782',
        type: 'object',
        calling: 'too many arguments',
        participants: 'SELECT DISTINCT IDENTITY_VAL_LOCAL() FROM %s;',
      },
    )
    expect(result).toMatchSnapshot()
  })

  test('2', () => {
    const result: any = mutations.default.setConversation(
      { calling: 'too many arguments', participants: 'SELECT * FROM %s;' },
      {
        id: '03ea49f8-1d96-4cd0-b279-0684e3eec3a9',
        type: 'object',
        calling: 'Message box: foo; bar\n',
        participants: 'INSERT INTO ',
      },
    )
    expect(result).toMatchSnapshot()
  })

  test('3', () => {
    const result: any = mutations.default.setConversation(
      { calling: 'error\n', participants: 'SELECT COUNT(*) FROM %s' },
      {
        id: '7289708e-b17a-477c-8a77-9ab575c4b4d8',
        type: 'number',
        calling: 'ValueError',
        participants: 'SELECT * FROM %s;',
      },
    )
    expect(result).toMatchSnapshot()
  })

  test('4', () => {
    const result: any = mutations.default.setConversation(
      {
        calling: 'Message box: foo; bar\n',
        participants: 'DELETE FROM %s WHERE expires < %%s',
      },
      {
        id: '03ea49f8-1d96-4cd0-b279-0684e3eec3a9',
        type: 'array',
        calling: 'multiple errors occurred',
        participants: 'SELECT COUNT(*) FROM %s',
      },
    )
    expect(result).toMatchSnapshot()
  })

  test('5', () => {
    const result: any = mutations.default.setConversation(
      { calling: '', participants: '' },
      { id: '', type: '', calling: '', participants: '' },
    )
    expect(result).toMatchSnapshot()
  })
})

describe('misc', () => {
  let initialState: ConversationState = {
    id: '',
    type: 'friend',
    calling: false,
    participants: [],
  }

  test('mutations.default.setCalling', () => {
    const state = { ...initialState }
    const argument = true

    mutations.default.setCalling(state, argument)
    expect(state.calling).toBeTruthy()
  })
  test('mutations.default.setCalling', () => {
    const state = { ...initialState }
    const argument = false

    mutations.default.setCalling(state, argument)
    expect(state.calling).toBeFalsy()
  })
  test('mutations.default.setParticipants', () => {
    const state = { ...initialState }
    const argument = false

    mutations.default.setParticipants(state, argument)
    expect(state.participants).toEqual([])
  })
  test('mutations.default.setParticipants', () => {
    const state = { ...initialState }
    const argument = {
      peerId: 'peerId',
      address: 'address',
      name: 'name',
      profilePicture: 'profilePicture',
      state: ConversationConnection.CONNECTED,
      activity: ConversationActivity.ACTIVE,
      updatedAt: 123,
    }

    mutations.default.setParticipants(state, [argument])
    expect(state.participants).toEqual([argument])
  })
  test('mutations.default.addParticipant', () => {
    const state = { ...initialState }
    const argument = {
      peerId: 'peerId',
      address: 'address',
      name: 'name',
      profilePicture: 'profilePicture',
      state: ConversationConnection.CONNECTED,
      activity: ConversationActivity.ACTIVE,
      updatedAt: 123,
    }
    const argument2 = {
      peerId: 'peerId2',
      address: 'address2',
      name: 'name2',
      profilePicture: 'profilePicture2',
      state: ConversationConnection.CONNECTED,
      activity: ConversationActivity.ACTIVE,
      updatedAt: 123,
    }

    mutations.default.setParticipants(state, [argument]) // Initial set
    mutations.default.addParticipant(state, argument2) // Append
    expect(state.participants).toEqual([argument, argument2])
  })
})
