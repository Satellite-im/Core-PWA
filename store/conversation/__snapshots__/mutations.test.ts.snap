// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mutations.default.setConversation 0 1`] = `undefined`;

exports[`mutations.default.setConversation 1 1`] = `undefined`;

exports[`mutations.default.setConversation 2 1`] = `undefined`;

exports[`mutations.default.setConversation 3 1`] = `undefined`;

exports[`mutations.default.setConversation 4 1`] = `undefined`;

exports[`mutations.default.setConversation 5 1`] = `undefined`;
